### DISCUSSION TITLE
best memory option for java

### DISCUSSION THREAD

**Discussion by paulobreim** [sepinf-inc/IPED#1132] [Q&A]
Considering a machine with SSDs, 2 processors with 16 threads each and a large amount of memory (384GB), what is the best option to use in the command line?
java -Xms30G -Xmx30G -jar iped.jar ......
java -jar iped.jar ......
or another?

**Comment by mcquadenc** [comment]
Hi Paulo, 

the documentation says:

> IPED used memory increases at processing time if you have dozens of CPU cores (1GB per logical CPU core is a rule of thumb) and at analysis time if your case is huge (for dozens of millions of files you will need a few dozens of GBs). Generally, it is best to let the JVM allocate memory by itself (by default it takes 1/4 of your physical RAM) instead setting it with -Xms/-Xmx, although setting half of your physical RAM is another good rule of thumb

When you say threads, do you mean logical cores!?

barring better judgment, iped allocates up to 1/4 of its memory by default.


with the amount of available memory you have, I think it would be a good alternative to use ramdisk.

**Reply by lfcnassif** [reply]
You can use this command to check the java defaults on your (Windows) system:
`java -XX:+PrintFlagsFinal -version | findstr HeapSize`

> java -Xms30G -Xmx30G -jar iped.jar ......

I would say those are good default values for most cases given your number of threads/logical cores. Remaining memory would be used as IO cache by your OS, or maybe you could create a ramdrive for temp disk as @mcquadenc suggested, leaving at least the same amount of memory given to the application as free memory/IO cache.

**Reply by lfcnassif** [reply]
I've just updated this in the wiki:
`on most systems java 8 default max heap size is half the RAM and java 9 and newer common default is 1/4 of the RAM`

**Comment by paulobreim** [comment]
Very good. 
Here is the result about java defaults:
c:\iped-3.18.14\iped-3.18.14>java -XX:+PrintFlagsFinal -version | findstr HeapSize
    uintx ErgoHeapSizeLimit                         = 0                                   {product}
    uintx HeapSizePerGCThread                       = 87241520                            {product}
    uintx InitialHeapSize                          := 2147483648                          {product}
    uintx LargePageHeapSizeThreshold                = 134217728                           {product}
    uintx MaxHeapSize                              := 2145386496                          {product}
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)


I will try use  ramdisk, very goog idea.

Thank you

**Reply by lfcnassif** [reply]
> uintx MaxHeapSize := 2145386496 {product}

See the default java max memory on your system is just 2GB, so you would be better using -Xmx30G

**Reply by paulobreim** [reply]
Thank you. I will do it.

**Comment by jojo-nc** [comment]
> Considering a machine with SSDs, 2 processors with 16 threads each and a large amount of memory (384GB), what is the best option to use in the command line? java -Xms30G -Xmx30G -jar iped.jar ...... java -jar iped.jar ...... or another?

Arguments `-Xms30G` and `-Xmx30G` (or any memory size you want to set) also work with the compiled iped.exe (like `-d input_file` or `-o output_dir`)
